File: Cpp/0918_maximum_sum_circular_subarray.cpp
// https://leetcode.cn/problems/maximum-sum-circular-subarray/

class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int n = nums.size();
        vector<int> cumsum(2*n, 0);
        for (int i = 0; i < 2*n-1; ++i) {
            cumsum[i+1] = cumsum[i] + nums[i%n];
        }
        auto comp = [&](int i, int j) {
            return cumsum[i] > cumsum[j];
        };
        // Use a priority queue to save the indices such that
        // cumsum[pq.top()] <= cumsum[k] for any k in pq.
        priority_queue<int, vector<int>, decltype(comp)> pq(comp);
        for (int i = 0; i < n; ++i) {
            pq.push(i);
        }
        int res = cumsum[n] - cumsum[pq.top()];
        for (int i = n; i < 2*n-1; ++i) {
            pq.push(i);
            while (pq.top() < i+1-n) {
                pq.pop();
            }
            res = max(res, cumsum[i+1] - cumsum[pq.top()]);
        }
        return res;
    }
};

File: Python3/0918_maximum_sum_circular_subarray.py
# https://leetcode.cn/problems/maximum-sum-circular-subarray/
import heapq
class Solution:
    def maxSubarraySumCircular(self, nums):
        """
        Inputs:
            nums: list[int]
        Outputs:
            res: int
        """
        n = len(nums)
        cumsum = [0 for _ in range(2*n)]
        for i in range(2*n-1):
            cumsum[i+1] = cumsum[i] + nums[i%n]
        # Make a priority queue pq that save indices and that satisfies
        # cumsum[pq[0]] <= cumsum[pq[k]] for any k in pq.
        pq = []
        for i in range(n):
            heapq.heappush(pq, (cumsum[i], i))
        res = cumsum[n] - cumsum[pq[0][1]]
        for i in range(n, 2*n-1):
            heapq.heappush(pq, (cumsum[i], i))
            while pq[0][1] < i+1-n:
                heapq.heappop(pq)
            res = max(res, cumsum[i+1] - cumsum[pq[0][1]])
        return res

sol = Solution()
nums = [1, -2, 3, -2]
print(sol.maxSubarraySumCircular(nums))